tests/components/apcupsd/test_init.py

```diff
@@ -28,74 +29,37 @@
     ],
 )
-async def test_async_setup_entry(hass: HomeAssistant, status: OrderedDict) -> None:
-    """Test a successful setup entry."""
-    # Minimal status does not contain "SERIALNO" field, which is used to determine the
-    # unique ID of this integration. But, the integration should work fine without it.
-    # In such a case, the device will not be added either
-    await async_init_integration(hass, status=status)
-
-    prefix = ""
-    if "SERIALNO" in status:
-        prefix = slugify(status.get("UPSNAME", "APC UPS")) + "_"
-
-    # Verify successful setup by querying the status sensor.
-    state = hass.states.get(f"binary_sensor.{prefix}online_status")
-    assert state
-    assert state.state != STATE_UNAVAILABLE
-    assert state.state == "on"
-
-
-@pytest.mark.parametrize(
-    "status",
-    [
-        # We should not create device entries if SERIALNO is not reported.
-        MOCK_MINIMAL_STATUS,
-        # We should set the device name to be the friendly UPSNAME field if available.
-        MOCK_MINIMAL_STATUS | {"SERIALNO": "XXXX", "UPSNAME": "MyUPS"},
-        # Otherwise, we should fall back to default device name --- "APC UPS".
-        MOCK_MINIMAL_STATUS | {"SERIALNO": "XXXX"},
-        # We should create all fields of the device entry if they are available.
-        MOCK_STATUS,
-    ],
-)
-async def test_device_entry(
-    hass: HomeAssistant, status: OrderedDict, device_registry: dr.DeviceRegistry
+async def test_async_setup_entry(
+    hass: HomeAssistant,
+    status: OrderedDict,
+    device_registry: dr.DeviceRegistry,
+    snapshot: SnapshotAssertion,
 ) -> None:
-    """Test successful setup of device entries."""
+    """Test a successful setup entry."""
     await async_init_integration(hass, status=status)
 
-    # Verify device info is properly set up.
-    if "SERIALNO" not in status:
-        assert len(device_registry.devices) == 0
-        return
-
-    assert len(device_registry.devices) == 1
-    entry = device_registry.async_get_device({(DOMAIN, status["SERIALNO"])})
-    assert entry is not None
-    # Specify the mapping between field name and the expected fields in device entry.
-    fields = {
-        "UPSNAME": entry.name,
-        "MODEL": entry.model,
-        "VERSION": entry.sw_version,
-        "FIRMWARE": entry.hw_version,
-    }
-
-    for field, entry_value in fields.items():
-        if field in status:
-            assert entry_value == status[field]
-        # Even if UPSNAME is not available, we must fall back to default "APC UPS".
-        elif field == "UPSNAME":
-            assert entry_value == "APC UPS"
-        else:
-            assert not entry_value
-
-    assert entry.manufacturer == "APC"
+    config_entry = await async_init_integration(hass, status=status)
+    device_entry = device_registry.async_get_device(
+        identifiers={(DOMAIN, config_entry.unique_id)}
+    )
+    device_name = "None" if device_entry is None else device_entry.name
```

Comment:
Why?

Reply:
When `UPSNAME` and `SERIALNO` are not present in the reported status, we currently do not create a device (where we have a mock status to test that)

Perhaps it's not appropriate to test `None` via snapshot, I've updated the logic here. PTAL

I remember we've discussed in the past that we can use the config entry ID as the unique ID. We had some handling for it but it's not working properly yet. I'll create a follow-up PR to fix that (as well as the tests).

But do let me know if you prefer we fix that directly in this PR.

Reply:
Ideally we don't have if statements in the tests. But we should then split this test case into 2, 1 where we test with the device and another one that confirms that it keeps working with that other scenario

Reply:
Yeah I definitely agree this is not nice, but since we are going to fix that (and remove the separate test shortly) in a follow-up PR, perhaps it's not that worth it?

I don't have strong preferences either way - I can create a follow-up and we can merge them together, or to keep this self-contained we can separate the tests and remove it later 🙂

Reply:
I'd rather have no if statements in here, also, is the change just a code wise change or a bugfix? Because in the latter case I would like to merge that first so it can go in a hotfix

Reply:
Sounds good!

The change is a bugfix - #130852, could you review and merge that one? I'll rebase this one once that gets in.